source/core/DualTypeConversion.cpp

```diff
@@ -0,0 +1,316 @@
+                ApplyNumeric<Int8>(int8ValueX, typeRefY, pDataY);
+                break;
+            }
+            case 2: {
+                Int16 int16ValueX = *static_cast<Int16*>(pDataX);
+                ApplyNumeric<Int16>(int16ValueX, typeRefY, pDataY);
+                break;
+            }
+            case 4: {
+                Int32 int32ValueX = *static_cast<Int32*>(pDataX);
+                ApplyNumeric<Int32>(int32ValueX, typeRefY, pDataY);
+                break;
+            }
+            case 8: {
+                Int64 int64ValueX = *static_cast<Int64*>(pDataX);
+                ApplyNumeric<Int64>(int64ValueX, typeRefY, pDataY);
+                break;
+            }
+        }
+        return true;
+    }
+
+    //------------------------------------------------------------
+    bool DualTypeConversion::ApplyIEEE754Binaries(TypeRef typeRefX, void* pDataX, TypeRef typeRefY, void* pDataY)
+    {
+        if (typeRefX->TopAQSize() == sizeof(Single)) {
+            Single singleValueX = *reinterpret_cast<Single*>(pDataX);
+            ApplyNumeric<Single>(singleValueX, typeRefY, pDataY);
+        } else {
+            Double doubleValueX = *reinterpret_cast<Double*>(pDataX);
+            ApplyNumeric<Double>(doubleValueX, typeRefY, pDataY);
+        }
+        return true;
+    }
+
+    //------------------------------------------------------------
+    bool DualTypeConversion::AreBooleanCompatible(TypeRef typeRefX, TypeRef typeRefY)
+    {
+        bool sameEncodingAndSize = DoTypesHaveSameEncodingAndSize(typeRefX, typeRefY);
+        return sameEncodingAndSize;
+    }
+
+    //------------------------------------------------------------
+    bool DualTypeConversion::AreUIntCompatible(TypeRef typeRefX, TypeRef typeRefY)
+    {
+        bool typesAreCompatible = TypesAreCompatible(typeRefX, typeRefY);
+        return typesAreCompatible;
+    }
+
+    //------------------------------------------------------------
+    bool DualTypeConversion::AreS2CIntCompatible(TypeRef typeRefX, TypeRef typeRefY)
+    {
+        bool typesAreCompatible = TypesAreCompatible(typeRefX, typeRefY);
+        return typesAreCompatible;
+    }
+
+    //------------------------------------------------------------
+    bool DualTypeConversion::AreIEEE754BinaryCompatible(TypeRef typeRefX, TypeRef typeRefY)
+    {
+        bool typesAreCompatible = TypesAreCompatible(typeRefX, typeRefY);
+        return typesAreCompatible;
+    }
+
+    //------------------------------------------------------------
+    bool DualTypeConversion::AreIntrinsicClustersCompatible(TypeRef typeRefX, TypeRef typeRefY)
+    {
+        if (typeRefX->IsComplex() && typeRefY->IsComplex()) {
+            return true;
+        }
+        SubString typeXName, typeYName;
+        Boolean isTypeXIntrinsicClusterType = typeRefX->IsIntrinsicClusterDataType(&typeXName);
+        Boolean isTypeYIntrinsicClusterType = typeRefY->IsIntrinsicClusterDataType(&typeYName);
```

Comment:
Don't we need to use isTypeXIntrinsicClusterType and isTypeYIntrinsicClusterType?

Reply:
Not in this case. The call to `IsIntrinsicClusterDataType` is necessary to get `typeXName` and `typeYName` out.